basic_subscriber.go

```diff
@@ -46,6 +55,48 @@ func (s *BasicSubscriber) InsertData(sd *StreamDataMsg) error {
 
 //Subscribe kicks off a go routine that calls the gotData func for every new video chunk
 func (s *BasicSubscriber) Subscribe(ctx context.Context, gotData func(seqNo uint64, data []byte, eof bool)) error {
+	glog.Infof("%v Sending SubReq message", s.Network.NetworkNode.ID())
+	return s.sendSub(ctx, SubReqID, SubReqMsg{StrmID: s.StrmID}, gotData)
+}
+
+func (s *BasicSubscriber) TranscoderSubscribe(ctx context.Context, gotData func(seqNo uint64, data []byte, eof bool)) error {
+	ts := TranscodeSubMsg{
+		MultiAddrs: s.Network.NetworkNode.Host().Addrs(),
+		NodeID:     s.Network.NetworkNode.ID(),
+		StrmID:     s.StrmID,
+	}
+	sig, err := s.Network.NetworkNode.Sign(ts.BytesForSigning())
+	if err != nil {
+		glog.Errorf("Error signing TranscodeSubMsg: %v", err)
+		return err
+	}
+	ts.Sig = sig
+	s.Network.NetworkNode.Host().Network().Notify(s)
+	glog.Infof("%v Sending TranscodeSub message for %v", s.Network.NetworkNode.ID(), s.StrmID)
+
+	// periodic retransmit in case the broadcaster misses the first message(s)
```

Comment:
Is re-transmit the proper strategy here?  What would make a broadcaster miss the message the first time, and get it in the future?  In the current implementation, when will sendCtx.Done() be triggered?

ps - I've been using https://github.com/cenkalti/backoff for a lot of the retry stuff for Infura websockets, might save you some cruft code.

Reply:
> What would make a broadcaster miss the message the first time, and get it in the future?

With the current peer routing mechanism, we have no feedback on delivery or message propagation, and no visibility into the network's connectivity. There could be intermittent connectivity anywhere along the route, and retrying gives the topology time to heal and work around routing partitions (especially since we only forward to one peer at a time).

While hopefully this issue wont be too common, we do want to be prepared when it happens; otherwise it'd be extremely difficult to diagnose, and frustrating for a user who just spent gas on a job transaction.

> Is re-transmit the proper strategy here?

What are the other options?

> will sendCtx.Done() be triggered?

Three possible points/scenarios where sendCtx.Done would be triggered:

* After the direct connection is established. This is an indicator that everything went normally.
* After receiving a data message. If no direct connection was received, this usually means the transcoder is unreachable from the broadcaster.
* If the subscribe context gets cancelled for whatever reason, before the first two cases above.

>  I've been using https://github.com/cenkalti/backoff for a lot of the retry stuff for Infura websockets

Looked into it briefly, but I'm not keen on pulling an entire dependency just to save a few lines. Maybe if we have to re-use this logic later.